### PR TITLE
Let dead code removal handle removing more of the unneeded code after prange conversion to parfor.

### LINKED ISSUE #4569: Error producing ndarray from parallel list comprehension
Error producing ndarray from parallel list comprehension--works fine serially with ndarrays, works parallel when returning list, both break it:

## Minimal reproducer:

```
import numpy as np
from numba import njit, prange

@njit(parallel=True)
def get_lengths_as_list_par(xx): # works
    return [len(xx[i]) for i in prange(len(xx))]

@njit(parallel=False)
def get_lengths_as_ndarray_serial(xx): # works
    return np.array([len(xx[i]) for i in range(len(xx))])

@njit(parallel=True)
def get_lengths_as_ndarray_par(xx): # breaks
    return np.array([len(xx[i]) for i in prange(len(xx))])

xx = [np.array([1,2,3], dtype=int),np.array([1,2], dtype=int)]

get_lengths_as_list_par(xx) ## works fine
get_lengths_as_ndarray_serial(xx) ## works fine
get_lengths_as_ndarray_par(xx) ## produces below error
```
## Software versions

- numba: 0.42.1
- numpy: 1.16.2
- python: 3.6.8 

## Traceback:
```
KeyError                            Traceback (most recent call last)
/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    616     try:
--> 617         yield
    618     except NumbaError as e:

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in lower_block(self, block)
    258                                    loc=self.loc, errcls_=defaulterrcls):
--> 259                 self.lower_inst(inst)
    260 

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in lower_inst(self, inst)
    307             ty = self.typeof(inst.target.name)
--> 308             val = self.lower_assign(ty, inst)
    309             self.storevar(val, inst.target.name)

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in lower_assign(self, ty, inst)
    453         elif isinstance(value, ir.Expr):
--> 454             return self.lower_expr(ty, value)
    455 

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in lower_expr(self, resty, expr)
    900         elif expr.op in ('getiter', 'iternext'):
--> 901             val = self.loadvar(expr.value.name)
    902             ty = self.typeof(expr.value.name)

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in loadvar(self, name)
   1070         """
-> 1071         ptr = self.getvar(name)
   1072         return self.builder.load(ptr)

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in getvar(self, name)
   1064         """
-> 1065         return self.varmap[name]
   1066 

KeyError: '$0.12'

During handling of the above exception, another exception occurred:

LoweringError                       Traceback (most recent call last)
<ipython-input-50-6d1e380882d3> in <module>()
     16 get_lengths_as_list_par(xx) ## works fine
     17 get_lengths_as_ndarray_serial(xx) ## works fine
---> 18 get_lengths_as_ndarray_par(xx) ## works fine

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    365                     e.patch_message(''.join(e.args) + help_msg)
    366             # ignore the FULL_TRACEBACKS config, this needs reporting!
--> 367             raise e
    368 
    369     def inspect_llvm(self, signature=None):

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    322                 argtypes.append(self.typeof_pyval(a))
    323         try:
--> 324             return self.compile(tuple(argtypes))
    325         except errors.TypingError as e:
    326             # Intercept typing error that may be due to an argument

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler_lock.py in _acquire_compile_lock(*args, **kwargs)
     30         def _acquire_compile_lock(*args, **kwargs):
     31             with self:
---> 32                 return func(*args, **kwargs)
     33         return _acquire_compile_lock
     34 

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/dispatcher.py in compile(self, sig)
    653 
    654             self._cache_misses[sig] += 1
--> 655             cres = self._compiler.compile(args, return_type)
    656             self.add_overload(cres)
    657             self._cache.save_overload(sig, cres)

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/dispatcher.py in compile(self, args, return_type)
     80                                       args=args, return_type=return_type,
     81                                       flags=flags, locals=self.locals,
---> 82                                       pipeline_class=self.pipeline_class)
     83         # Check typing error if object mode is used
     84         if cres.typing_error is not None and not flags.enable_pyobject:

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library, pipeline_class)
    924     pipeline = pipeline_class(typingctx, targetctx, library,
    925                               args, return_type, flags, locals)
--> 926     return pipeline.compile_extra(func)
    927 
    928 

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in compile_extra(self, func)
    372         self.lifted = ()
    373         self.lifted_from = None
--> 374         return self._compile_bytecode()
    375 
    376     def compile_ir(self, func_ir, lifted=(), lifted_from=None):

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in _compile_bytecode(self)
    855         """
    856         assert self.func_ir is None
--> 857         return self._compile_core()
    858 
    859     def _compile_ir(self):

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in _compile_core(self)
    842         self.define_pipelines(pm)
    843         pm.finalize()
--> 844         res = pm.run(self.status)
    845         if res is not None:
    846             # Early pipeline completion

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler_lock.py in _acquire_compile_lock(*args, **kwargs)
     30         def _acquire_compile_lock(*args, **kwargs):
     31             with self:
---> 32                 return func(*args, **kwargs)
     33         return _acquire_compile_lock
     34 

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in run(self, status)
    253                     # No more fallback pipelines?
    254                     if is_final_pipeline:
--> 255                         raise patched_exception
    256                     # Go to next fallback pipeline
    257                     else:

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in run(self, status)
    244                 try:
    245                     event(stage_name)
--> 246                     stage()
    247                 except _EarlyPipelineCompletion as e:
    248                     return e.result

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in stage_nopython_backend(self)
    715         """
    716         lowerfn = self.backend_nopython_mode
--> 717         self._backend(lowerfn, objectmode=False)
    718 
    719     def stage_compile_interp_mode(self):

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in _backend(self, lowerfn, objectmode)
    664             self.library.enable_object_caching()
    665 
--> 666         lowered = lowerfn()
    667         signature = typing.signature(self.return_type, *self.args)
    668         self.cr = compile_result(

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in backend_nopython_mode(self)
    651                 self.calltypes,
    652                 self.flags,
--> 653                 self.metadata)
    654 
    655     def _backend(self, lowerfn, objectmode):

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/compiler.py in native_lowering_stage(targetctx, library, interp, typemap, restype, calltypes, flags, metadata)
   1046         lower = lowering.Lower(targetctx, library, fndesc, interp,
   1047                                metadata=metadata)
-> 1048         lower.lower()
   1049         if not flags.no_cpython_wrapper:
   1050             lower.create_cpython_wrapper(flags.release_gil)

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in lower(self)
    176         if self.generator_info is None:
    177             self.genlower = None
--> 178             self.lower_normal_function(self.fndesc)
    179         else:
    180             self.genlower = self.GeneratorLower(self)

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in lower_normal_function(self, fndesc)
    217         # Init argument values
    218         self.extract_function_arguments()
--> 219         entry_block_tail = self.lower_function_body()
    220 
    221         # Close tail of entry block

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in lower_function_body(self)
    242             bb = self.blkmap[offset]
    243             self.builder.position_at_end(bb)
--> 244             self.lower_block(block)
    245 
    246         self.post_lower()

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/lowering.py in lower_block(self, block)
    257             with new_error_context('lowering "{inst}" at {loc}', inst=inst,
    258                                    loc=self.loc, errcls_=defaulterrcls):
--> 259                 self.lower_inst(inst)
    260 
    261     def create_cpython_wrapper(self, release_gil=False):

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/contextlib.py in __exit__(self, type, value, traceback)
     97                 value = type()
     98             try:
---> 99                 self.gen.throw(type, value, traceback)
    100             except StopIteration as exc:
    101                 # Suppress StopIteration *unless* it's the same exception that

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    623         from numba import config
    624         tb = sys.exc_info()[2] if config.FULL_TRACEBACKS else None
--> 625         six.reraise(type(newerr), newerr, tb)
    626 
    627 

/projects/ribo7412/software/anaconda/envs/fwd/lib/python3.6/site-packages/numba/six.py in reraise(tp, value, tb)
    657         if value.__traceback__ is not tb:
    658             raise value.with_traceback(tb)
--> 659         raise value
    660 
    661 else:

LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
'$0.12'

File "<ipython-input-50-6d1e380882d3>", line 12:
def get_lengths_as_ndarray_par(xx):
    return np.array([len(xx[i]) for i in prange(len(xx))])
    ^

[1] During: lowering "$implicit0.3549.3583 = getiter(value=$0.12)" at <ipython-input-50-6d1e380882d3> (12)
-------------------------------------------------------------------------------
This should not have happened, a problem has occurred in Numba's internals.
```

### PR DESCRIPTION
**Pull request by DrTodd13**
Resolves #4569.

The code to convert pranges to parfors was previously removing the prange call and the loop header block under the assumption that those statements would not be used anywhere else but this turns out not to always be the case.  In this PR, I leave the prange call in there and move the header block into the first loop body block and then let dead code elimination remove statements that aren't needed.

### CODE REVIEW

**Comment by sklam** [comment]
Test seems to be hanging at `numba.tests.test_parfors.TestParfors.test_logistic_regression`

File: numba/parfor.py [hunk 1]
```diff
@@ -1657,13 +1657,26 @@ def _convert_loop(self, blocks):
...
                                 and isinstance(stmt.value, ir.Expr)
                                 and stmt.value.op == 'pair_first'):
                             loop_index = stmt.target.name
+                            li_index = hbi
                             break
+                    assert(loop_index is not None)
+                    # Remove pair_first from header.
+                    # We have to remove the pair_first by hand since it causes problems
+                    # for some code below if we don't.
+                    header_body = header_body[:li_index] + header_body[li_index+1:]
```
Comment (stuartarchibald) [inline numba/parfor.py]: would it be a good idea to check that the thing being sliced out is actually a `pair_first`?

Comment (DrTodd13) [reply numba/parfor.py]: The code above this assures that that is indeed the case.  We search for pair_first and loop_index gets set only if it is found then after the loop we assert it is found.  Thus, li_index guaranteed to be that of a pair_first.

Comment (stuartarchibald) [reply numba/parfor.py]: Ah I see, thanks.

**Comment by stuartarchibald** [comment]
> Thanks for the patch. I think that it'd be good to add at least the original failing code in as a test case to demonstrate that this new code path does indeed fix the issue. Thanks.

^ This still needs resolving.

**Comment by DrTodd13** [comment]
> > Thanks for the patch. I think that it'd be good to add at least the original failing code in as a test case to demonstrate that this new code path does indeed fix the issue. Thanks.
> 
> ^ This still needs resolving.

Fixed now.

**Comment by stuartarchibald** [comment]
This is failing CI, seems like its reading junk from memory into list slots, not sure why? Will give it a rerun just to double check.

**Comment by DrTodd13** [comment]
These failures don't happen when run individually.  What do do?  @stuartarchibald

**Comment by stuartarchibald** [comment]
I think the git history for this PR has gone wrong somewhere? There's now a lot of patches!?

**Comment by stuartarchibald** [comment]
> These failures don't happen when run individually. What do do? @stuartarchibald

If you checkout master and run the full test suite, do you get any failures? If not then that probably limits the scope of the issue to new code in this PR.

**Comment by stuartarchibald** [comment]
Closing, https://github.com/numba/numba/pull/4991 carries on.

**Pull request closed without merging**